fetcher/fedora.go

```diff
@@ -0,0 +1,293 @@
+	const moduleHref = "https://dl.fedoraproject.org/pub/fedora/linux/updates/%s/Modular/x86_64/repodata/repomd.xml"
+	for _, v := range target {
+		reqs = append(reqs, fetchRequest{
+			target:       v,
+			url:          fmt.Sprintf(href, v),
+			mimeType:     mimeTypeXml,
+			concurrently: false,
+		})
+		moduleReqs = append(moduleReqs, fetchRequest{
+			target:       v,
+			url:          fmt.Sprintf(moduleHref, v),
+			mimeType:     mimeTypeXml,
+			concurrently: false,
+		})
+	}
+	return
+}
+
+func fetchEverythingFedora(reqs []fetchRequest) (fedoraUpdatesPerVersion, error) {
+	log15.Info("start fetch data from Everything/x86_64/repodata/repomd.xml")
+	feeds, err := fetchFeedFilesFedora(reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	updates, err := fetchUpdateInfosFedora(feeds)
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := parseFetchResultsFedora(updates)
+	if err != nil {
+		return nil, err
+	}
+
+	for version, v := range results {
+		log15.Info(fmt.Sprintf("%d CVEs for Fedora %s Fetched", len(v.UpdateList), version))
+	}
+
+	return results, nil
+}
+
+func fetchModulesFedora(reqs []fetchRequest) (fedoraUpdatesPerVersion, error) {
+	log15.Info("start fetch data from Modular/x86_64/repodata/repomd.xml")
+	feeds, err := fetchModuleFeedFilesFedora(reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	updates, err := fetchUpdateInfosFedora(feeds)
+	if err != nil {
+		return nil, err
+	}
+
+	moduleYaml, err := fetchModulesYamlFedora(feeds)
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := parseFetchResultsFedora(updates)
+	if err != nil {
+		return nil, err
+	}
+
+	for version, result := range results {
+		log15.Info(fmt.Sprintf("%d CVEs for Fedora %s modules Fetched", len(result.UpdateList), version))
+		for i, update := range result.UpdateList {
+			yml, ok := moduleYaml[version][update.Title]
+			if ok {
+				var pkgs []Package
+				for _, rpm := range yml.Data.Artifacts.Rpms {
+					pkg, err := rpm.NewPackageFromRpm()
+					if err != nil {
+						return nil, err
+					}
+					pkgs = append(pkgs, pkg)
+				}
+				results[version].UpdateList[i].Packages = pkgs
+				results[version].UpdateList[i].ModularityLabel = yml.ConvertToModularityLabel()
+			}
+		}
+	}
+	return results, nil
+}
+
+func fetchFeedFilesFedora(reqs []fetchRequest) ([]FetchResult, error) {
+	if len(reqs) == 0 {
+		return nil, fmt.Errorf("There are no versions to fetch")
+	}
+	results, err := fetchFeedFiles(reqs)
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to fetch. err: %w", err)
+	}
+	return results, nil
+}
+
+func fetchUpdateInfosFedora(results []FetchResult) ([]FetchResult, error) {
+	log15.Info("start fetch updateinfo in repomd.xml")
+	var updateInfoReqs []fetchRequest
+	for _, r := range results {
+		var repoMd RepoMd
+		if err := xml.NewDecoder(bytes.NewBuffer(r.Body)).Decode(&repoMd); err != nil {
+			log15.Warn(fmt.Sprintf("Failed to decode repomd. Skip to fetch version %s", r.Target), "err", err)
+			continue
+		}
```

Comment:
That's what Amazon does, but can we go through the case where the decode of repomd.xml fails?